UserService.groovy

```diff
@@ -0,0 +1,236 @@
+            smartPlaylistService.deleteAllSmartPlaylists(user)
+            // TODO Only for development, remove before production
+
+            smartPlaylistService.createDefaultSmartPlaylists(user)
+        }
+        user.save(flush: true)
+        return user
+    }
+
+    User createUser(Profile profile) {
+        User user = new User(profile)
+        user.save(flush: true)
+        log.info "Account created for $user.username:$user.displayName"
+        return user
+    }
+
+    Profile getProfile(String accessToken) {
+        return apiService.getProfile(accessToken)
+    }
+
+    void logon(User user) {
+        user.status = UserStatus.ACTIVE
+        user.lastChanged = new Date()
+        user.save(flush: true)
+        userInfoServiceProxy.id = user.id
+        log.info "${user?.displayName} has logged on"
+    }
+
+    void logoff(User user) {
+        userInfoServiceProxy.id = null
+        log.info "${user?.displayName} has logged off"
+    }
+
+    void deactivate(User user) {
+        user.status = UserStatus.DEACTIVATED
+        user.lastRecorded?.delete()
+        user.currentlyPlaying?.delete()
+        user.lastRecorded = null
+        user.currentlyPlaying = null
+        user.save(flush: true)
+        log.info "${user?.displayName} has deactivated their account"
+    }
+
+    void reactivate(User user) {
+        user.status = UserStatus.ACTIVE
+        user.lastChanged = new Date()
+        user.save(flush: true)
+        log.info "${user?.displayName} has reactivated their account"
+    }
+
+    void delete(User user) {
+        List<SmartPlaylist> playlists = SmartPlaylist.findAllByUser(user)
+        List<TrackHistory> history = TrackHistory.findAllByUser(user)
+
+        playlists.each { playlist ->
+            smartPlaylistService.deleteSmartPlaylist(user, playlist.id)
+        }
+
+        TrackHistory.deleteAll(history)
+
+        LastRecorded.findByUser(user)?.delete()
+        CurrentlyPlaying.findByUser(user)?.delete()
+        Authentication.findByUser(user)?.delete()
+        user.delete(flush: true)
+        log.info "${user?.displayName} deleted their account"
+    }
+
+    Authentication getAuthentication(String code, String state) {
+        if (state != this.state && state != null) {
+            throw new AuthenticationException('Invalid state', 'The server did not return the same state that it was supplied with')
+        }
+        log.debug("Authorisation code: $code")
+        log.debug("state received: $state")
+        this.state = null
+        Token token = apiService.getToken(code)
+        return token.authentication
+    }
+
+    void setAuthentication(User user, Authentication authentication) {
+        Authentication auth = Authentication.findOrCreateByUser(user)
+        auth.with {
+            accessToken = authentication.accessToken
+            refreshToken = authentication.refreshToken
+            expiry = authentication.expiry
+        }
```

Comment:
Adding this comment as part of a review (hint: "Pending" means no-one else can see it!)